src/dos/drives.cpp

```diff
@@ -311,6 +327,248 @@ char *DriveManager::GetDrivePosition(int drive)
 	return swap_position;
 }
 
+void DriveManager::SetDiskSpeed(DiskSpeed disk_speed, DiskType disk_type)
+{
+	switch (disk_type) {
+	case DiskType::Floppy: disk_settings.fdd_disk_speed = disk_speed; break;
+	case DiskType::HardDisk:
+		disk_settings.hdd_disk_speed = disk_speed;
+		break;
+	case DiskType::CdRom:
+		disk_settings.cdrom_disk_speed = disk_speed;
+		break;
+	default:
+		LOG_WARNING("DriveManager: Unknown disk type %d",
+		            static_cast<int>(disk_type));
+		return;
+	}
+}
+
+void DriveManager::RegisterIoCallback(std::function<void()> callback, DiskType disk_type)
+{
+	switch (disk_type) {
+	case DiskType::Floppy: io_callback_floppy = std::move(callback); break;
+	case DiskType::HardDisk:
+		io_callback_harddisk = std::move(callback);
+		break;
+	case DiskType::CdRom: io_callback_cdrom = std::move(callback); break;
+	default:
+		LOG_WARNING("DriveManager: Unknown disk type %d",
+		            static_cast<int>(disk_type));
+		return;
+	}
+}
+
+// TODO: Unregister should work on particular callbacks instead of all of a
+// given disk type
+void DriveManager::UnregisterIoCallback(DiskType disk_type)
+{
+	switch (disk_type) {
+	case DiskType::Floppy: io_callback_floppy.reset(); break;
+	case DiskType::HardDisk: io_callback_harddisk.reset(); break;
+	case DiskType::CdRom: io_callback_cdrom.reset(); break;
+	default: break;
+	}
+}
+
+// Call any registered callbacks in the supplied callback vector
+void DriveManager::ExecuteRegisteredCallbacks(DiskType disk_type)
+{
+	switch (disk_type) {
+	case DiskType::Floppy:
+		if (io_callback_floppy.has_value()) {
```

Comment:
Just `if (io_callback_floppy) { ... }` (drop `has_value()`).
Same comment for the rest of the file; please drop `has_value()` everywhere.